flink-runtime/src/main/java/org/apache/flink/runtime/resourcemanager/slotmanager/FineGrainedSlotManager.java

```diff
@@ -594,18 +597,19 @@ private void checkResourceRequirementsWithDelay() {
         if (requirementsCheckDelay.toMillis() <= 0) {
             checkResourceRequirements();
         } else {
-            if (requirementsCheckFuture == null || requirementsCheckFuture.isDone()) {
-                requirementsCheckFuture = new CompletableFuture<>();
-                scheduledExecutor.schedule(
-                        () ->
-                                mainThreadExecutor.execute(
-                                        () -> {
-                                            checkResourceRequirements();
-                                            Preconditions.checkNotNull(requirementsCheckFuture)
-                                                    .complete(null);
-                                        }),
-                        requirementsCheckDelay.toMillis(),
-                        TimeUnit.MILLISECONDS);
+            if (requirementsCheckFuture.isDone()) {
+                requirementsCheckFuture =
+                        scheduledExecutor.schedule(
+                                () -> {
+                                    if (started) {
```

Comment:
This read is unsafe since `started` isn't volatile.

Reply:
yeah, I had to revisit the PR. I looked into the synchronization of `started` yesterday and was kind of puzzled why we haven't had to synchronize the field till now. But that's due to the fact that all methods that rely on the `started` field are actually called from within the ResourceManager's main thread. In the end, I forgot to consider this when actually accessing the `SlotManager`'s state from another thread :facepalm: 

Anyway, I reiterated over the PR and added some assertions to make it clearer that the `FineGrainedSlotManager` should be handled from within the `ResourceManager`'s main thread. Please clarify if I came up with the wrong conclusion and missed something here.